Deprecating EoL DB versions, support for more DB versions

mollux opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 7.x)
* a.b for any bug fixes (e.g. 5.2, 6.0)
* c.x for any bug fixes, features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 7.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌ <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | ✔️
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️ <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #13262 <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

This PR aims to make Mautic officially compatible with newer Mysql and Mariadb versions.
This PR adds / changes to LTS versions of MariaDB and MySQL.

The main changes are related to sorting, as there is a change on how MySQL 8 approaches this.

When merged, this also needs to be adapted on the [website](https://mautic.org/mautic-requirements/#:~:text=Mautic%20supports%20MySQL%20and%20MariaDB%20out%20of%20the%20box.&text=Minimum%20MySQL%20version%20is%205.7%20or%20MariaDB%2010.2.) 

This PR can be rebased easily on 6.x (and most likely 5.x too if needed)

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!-- Remove HTML comment markup below to use the table for screenshots when relevant. -->
<!--
| Before                                 | After
| -------------------------------------- | ---
|                                        | 
-->


---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. 

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

mollux committed:
make tests work with LTS versions of MariaDB and MySQL

CS fix

no mariadb:11.4 as it breaks tests

change to correct db version to ensure tests are executed

changing the relevant versions only for Mautic 7 on

Let's see if this works

Let's see if this makes more sense

another try

no comment

no comment part 2

let's try a cleaner way

matbcvo commented on .github/workflows/tests.yml:
```diff
@@ -58,7 +57,7 @@ jobs:
 
     - name: PHPUnit cache
-      if: ${{matrix.php-versions == '8.2' && matrix.db-types == 'mariadb'}}
+      if: ${{matrix.php-versions == '8.2' && matrix.db-types == 'mariadb:11.4'}}
```
Use a version that exists in `matrix.db-types`, otherwise this step will never be executed. `mariadb:11.4` is not listed in `matrix.db-types`. The same applies to other steps where `mariadb:11.4` is used.

mollux replied:
right, that's a copy paste issue, fixing it!

RCheesley commented:
@escopecz  we can certainly ask if we get close to our limits regularly

mollux commented:
@escopecz we also could only target the latest per major version.
So that would mean `['mysql:5.7', 'mysql:8.4', 'mysql:9.2', 'mariadb:10.11']`

Not sure how fast we can drop support for MySQL 5.7. Maybe Mautic 7 should only support Mysql 8+ and Mariadb 10+, and we drop 5.7 in this PR entirely

mollux commented:
@escopecz another solution is to have a daily / weekly check that runs on 'not supported but probably used' versions, but that's not bringin any value IMO.

mollux commented:
I adopted the PR. Only version I want to add / check is mariadb 11.4 (LTS), as it makes senso to know if there are known broken functionalities

mollux commented:
Mariadb issue is related to https://mariadb.org/mariadb-server-docker-official-images-healthcheck-without-mysqladmin/, On it to fix it

mollux commented:
ok, all good, I squashed the commits as there was no real added value in the different steps.
All should be clear from the description and the changes themselves.

@escopecz I addressed your concern by limiting the number of SQL versions.
I don't think it's a good idea to skip testing for all php versions and all sql versions, as those are the actual combo's we support, even though it will be rare to hit issues there.

matbcvo reviewed (approved):
The code changes look good to me.

escopecz reviewed (commented):
I'm also fine with dropping MySql 5.7 as it is EoL for 2 years. Let's see if we'll be able to get enough processing time with GH actions to support this.

@mollux could you please mention the MySql 5.7 drop in https://github.com/mautic/mautic/blob/7.x/UPGRADE-7.0.md?

escopecz commented:
Looking through the code I can see we'll have to update release.yml, AppKernel.php, release_metadata.json and perhaps other places.

mollux commented:
@escopecz as there is now consensus on what we support and what not, I can make those changes

mollux committed:
Bump version requirement of Mysql of Mariadb

mollux commented:
@escopecz done, see https://github.com/mautic/mautic/pull/14836/commits/85e75dd694a5d2c4cea0ca690a39706768ef9393

matbcvo commented:
Please take a look at the failing E2E test.

```
Error:   - [error] Your database version (10.3.39-MariaDB-1:10.3.39+maria~ubu2004-log) is too old for Mautic to work correctly. Supported versions are MySQL 8.4.0 (or higher) and MariaDB 10.11.0 (or higher).
Install canceled
```

This can be fixed by making a change here:
https://github.com/mautic/mautic/blob/10c9235fd55b0f213679be0742918f5db400a0d4/.ddev/config.yaml#L7

mollux committed:
bump mariadb version in ddev config

mollux commented:
@matbcvo thanks :) well at least that proves the check works :)

escopecz reviewed (approved):
I see no issues. Thanks Mattias! We missed you!

@matbcvo can you please re-review also with the latest commits?

matbcvo reviewed (approved):
The code changes look good to me. 👍
